rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala

```diff
@@ -1,92 +1,57 @@
-  }
+  def emptyCost[F[_]: Sync]: F[Ref[F, CostState]] =
+    Ref.of[F, CostState](CostState(Cost(0, "init")))
 
-  def withPermit[F[_]: Sync, A](cost: _cost[F])(t: F[A]) =
-    Sync[F].bracket[Unit, A](cost.acquire)(kp(t))(kp(cost.release))
+  implicit class CostStateOps[F[_]](private val cost: CostStateRef[F]) extends AnyVal {
+    def set(amount: Cost): F[Unit] = cost.update(_.reset(amount))
 
-  def charge[F[_]: Sync](
-      amount: Cost
-  )(implicit cost: _cost[F], error: _error[F]): F[Unit] =
-    withPermit[F, Unit](cost)(
-      cost.get.flatMap { c =>
-        if (c.value < 0) error.raiseError[Unit](OutOfPhlogistonsError)
-        else cost.tell(Chain.one(amount)) >> cost.set(c - amount)
-      }
-    ) >> error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0).void
+    def current(implicit f: Functor[F]): F[Cost] = cost.get.map(_.total)
 
-  // TODO: Remove global (dummy) implicit!
-  implicit def noOpCostLog[M[_]: Applicative]: FunctorTell[M, Chain[Cost]] =
-    new DefaultFunctorTell[M, Chain[Cost]] {
-      override val functor: Functor[M]  = implicitly[Functor[M]]
-      def tell(l: Chain[Cost]): M[Unit] = Applicative[M].pure(())
-    }
-
-  implicit def ntCostLog[F[_]: Monad, G[_]: Sync](
-      nt: F ~> G
-  )(implicit C: _cost[F], mc: MonadCancel[F, Throwable]): _cost[G] =
-    new Semaphore[G] with MonadState[G, Cost] with FunctorTell[G, Chain[Cost]] {
-      override val functor: Functor[G]                  = Functor[G]
-      override def tell(l: Chain[Cost]): G[Unit]        = nt(C.tell(l))
-      override def tuple[A](ta: (Chain[Cost], A)): G[A] = nt(C.tuple(ta))
-
-      override val monad: Monad[G]                       = Monad[G]
-      override def writer[A](a: A, l: Chain[Cost]): G[A] = nt(C.writer(a, l))
-      override def get: G[Cost]                          = nt(C.get)
-      override def inspect[A](f: Cost => A): G[A]        = nt(C.inspect(f))
-      override def modify(f: Cost => Cost): G[Unit]      = nt(C.modify(f))
-      override def set(s: Cost): G[Unit]                 = nt(C.set(s))
+    // Used in tests
+    def <+(amount: Cost): F[Unit] = cost.update(s => s.copy(s.total + amount))
+  }
+}
 
-      override def acquireN(n: Long): G[Unit]       = nt(C.acquireN(n))
-      override def available: G[Long]               = nt(C.available)
-      override def count: G[Long]                   = nt(C.count)
-      override def releaseN(n: Long): G[Unit]       = nt(C.releaseN(n))
-      override def tryAcquireN(n: Long): G[Boolean] = nt(C.tryAcquireN(n))
-      override def permit: Resource[G, Unit]        = C.permit.mapK(nt)
-      override def mapK[K[_]](f: G ~> K)(implicit G: MonadCancel[K, _]): Semaphore[K] =
-        C.mapK(nt andThen f)
-    }
+package object accounting extends Costs {
 
+  def charge[F[_]: Sync: CostStateRef](amount: Cost): F[Unit] =
+    for {
+      isError <- CostStateRef[F].modify(_.charge(amount))
+      _       <- OutOfPhlogistonsError.raiseError.whenA(isError)
```

Comment:
its not required to fix since it might be a big change - but do we want OutOfPhlogistonsError at all to be thrown? or it should be a better practice for this function to return Either[Unit, Cost], where left means out of polo and right - phlo left after this charge. @tgrospic I recall this flow control though raising an error is not something we want.

Reply:
We use ApplicativeErrors to signal invalid behavior so it's fine to throw OutOfPhlogistonsError here.